gym/spaces/graph.py

```diff
@@ -0,0 +1,260 @@
+        """
+        self._np_random = None
+        if seed is not None:
+            if isinstance(seed, seeding.RandomNumberGenerator):
+                self._np_random = seed
+            else:
+                self.seed(seed)
+
+        assert isinstance(
+            node_space, (Box, Discrete)
+        ), "Values of the node_space should be instances of Box or Discrete"
+        if edge_space is not None:
+            assert isinstance(
+                edge_space, (Box, Discrete)
+            ), "Values of the edge_space should be instances of Box or Discrete"
+
+        self.node_space = node_space
+        self.edge_space = edge_space
+
+        # graph object creator
+        self.graphObj = GraphObj
+
+        super().__init__(None, None, seed)
+
+    def _generate_sample_space(self, base_space, num) -> Optional[Union[Box, Discrete]]:
+        # the possibility of this space having nothing
+        if num == 0:
+            return None
+
+        if isinstance(base_space, Box):
+            return Box(
+                low=np.array(max(1, num) * [base_space.low]),
+                high=np.array(max(1, num) * [base_space.high]),
+                shape=(num, *base_space.shape),
+                dtype=base_space.dtype,
+                seed=self._np_random,
+            )
+        elif isinstance(base_space, Discrete):
+            return MultiDiscrete(nvec=[base_space.n] * num, seed=self._np_random)
+        elif base_space is None:
+            return None
+        else:
+            raise AssertionError(
+                "Only Box and Discrete can be accepted as a base_space."
+            )
+
+    def _sample_sample_space(self, sample_space) -> Optional[np.ndarray]:
+        if sample_space is not None:
+            return sample_space.sample()
+        else:
+            return None
+
+    def sample(self) -> GraphObj:
+        """Returns a random sized graph space with num_nodes between 1 and 10."""
+        num_nodes = self.np_random.integers(low=1, high=10)
+
+        # we only have edges when we have at least 2 nodes
+        num_edges = 0
+        if num_nodes > 1:
+            # maximal number of edges is (n*n) allowing self connections and two way is allowed
+            num_edges = self.np_random.integers(num_nodes * num_nodes)
+
+        node_sample_space = self._generate_sample_space(self.node_space, num_nodes)
+        edge_sample_space = self._generate_sample_space(self.edge_space, num_edges)
+
+        sampled_nodes = self._sample_sample_space(node_sample_space)
+        sampled_edges = self._sample_sample_space(edge_sample_space)
+
+        sampled_node_list = None
+        if sampled_edges is not None:
+            if num_edges > 0:
+                sampled_node_list = self.np_random.integers(
+                    low=0, high=num_edges, size=(num_edges, 2)
+                )
+
+        return GraphObj(sampled_nodes, sampled_edges, sampled_node_list)
```

Comment:
Is this graph object necessary? Could we not return a tuple of nodes, adjacency matrix and sampled node list. 
Here, `sampled_node_list` and in graphobj `edge_links` have completely different variable names, why?

Reply:
The idea for using tuples can work, but also makes things much harder to enforce and perform checks for. But if that is desired, I can change it to be so.

`sampled_node_list` and `edge_links` is an artifact from some refactoring, thanks for catching this.